tests/factories.py

```diff
@@ -0,0 +1,144 @@
+    is_staff = False
+    is_active = True
+
+    # Use post-generation hook for setting password if needed for login tests
+    @factory.post_generation
+    def password(self, create, extracted, **kwargs):
+        if not create:
+            # Simple build, do nothing.
+            return
+        # Use a default password if none is provided
+        password = extracted or "default_password"
+        self.set_password(password)
+        self.save()  # Save again after setting password
+
+
+# --- Application Factory ---
+class ApplicationFactory(DjangoModelFactory):
+    """
+    Factory for oauth2_provider.models.Application.
+    Creates a confidential client application allowing authorization code grant by default.
+    """
+
+    class Meta:
+        model = Application
+
+    # Associate with a user (e.g., the owner/developer of the application)
+    # If you don't need a specific user owner, you can set user=None
+    # if your Application model allows it (check null=True on the user field).
+    # Usually, an Application requires a user.
+    user = factory.SubFactory(UserFactory)
+
+    # Sensible defaults for common use cases
+    client_type = Application.CLIENT_CONFIDENTIAL
+    authorization_grant_type = Application.GRANT_AUTHORIZATION_CODE
+
+    # Application name
+    name = factory.Sequence(lambda n: f"Test Application {n}")
+
+    # Redirect URIs - can be a space-separated string of URIs
+    # Often needed for authorization code flow.
+    redirect_uris = "https://example.com/callback http://localhost:8000/callback"
+
+    # client_id and client_secret are usually generated automatically by
+    # django-oauth-toolkit upon saving, so we don't typically define them here.
+    # skip_authorization = False # Default is False
+
+
+# --- AccessToken Factory ---
+class AccessTokenFactory(DjangoModelFactory):
+    """
+    Factory for oauth2_provider.models.AccessToken.
+    Creates an access token associated with a user and application.
+    """
+
+    class Meta:
+        model = AccessToken
+
+    # Link to a user (the resource owner granting access)
+    user = factory.SubFactory(UserFactory)
+
+    # Link to the client application the token is granted for
+    application = factory.SubFactory(ApplicationFactory)
+
+    # The token string itself - DOT usually generates this.
+    # If you need predictable tokens for testing, uncomment and use Sequence/LazyFunction.
+    # token = factory.Sequence(lambda n: f"test_token_{n}")
+    # Or use UUID for uniqueness:
+    token = factory.LazyFunction(lambda: uuid.uuid4().hex)
+
+    # Define the scope of the token (space-separated strings)
+    scope = "read write"  # Adjust as needed for your application's scopes
+
+    # Set an expiration date - typically slightly in the future
+    expires = factory.LazyFunction(lambda: timezone.now() + timedelta(hours=1))
+
+    # source_refresh_token = None # Optional: Link to a RefreshToken instance if needed
```

Comment:
And [AccessTokenFactory](https://github.com/openedx/edx-platform/blob/6db1e1db26a0d307446109334f49841aa9aae599/openedx/core/djangoapps/oauth_dispatch/tests/factories.py?plain=1#L36)